### PR TITLE
refact(locator): move some logic inside gateway

### PR DESCRIPTION
**Pull request by antmat**
This PR dedicates all logic for resoving a service to the gateway.
Thus gateway now is a mandatory component for locator.
API of the gateway is refactored to make it possible to resolve local services.
Cluster component subscriptions now are managed by config value, not the presence of the gateway component

### CODE REVIEW

**Comment by antmat** [comment]
It's WIP in fact, I think I will add more changes

**Comment by antmat** [comment]
updated description

File: include/cocaine/api/gateway.hpp [hunk 1]
```diff
@@ -30,25 +31,61 @@ namespace cocaine { namespace api {
 struct gateway_t {
     typedef gateway_t category_type;
 
+    struct service_description_t {
```
Comment (3Hren) [inline include/cocaine/api/gateway.hpp]: Why'd you ever need such abstraction?

**Comment by antmat** [comment]
First of all it is bound with https://github.com/3Hren/cocaine-plugins/pull/83 .

File: src/cluster/multicast.cpp [hunk 2]
```diff
@@ -136,12 +136,14 @@ multicast_t::multicast_t(context_t& context, interface& locator, const std::stri
 
     m_socket.set_option(multicast::join_group(m_cfg.endpoint.address()));
 
-    const auto announce = std::make_shared<announce_t>();
+    if(args.as_object().at("aggregate_node", false).as_bool()) {
+        const auto announce = std::make_shared<announce_t>();
 
-    m_socket.async_receive_from(buffer(announce->buffer.data(), announce->buffer.size()),
```
Comment (3Hren) [inline src/cluster/multicast.cpp]: Do not change code style without reason.

File: src/gateway/adhoc.cpp [hunk 3]
```diff
@@ -42,66 +56,109 @@ adhoc_t::~adhoc_t() {
...
-        {"uuid", lb->second.uuid}
-    }));
-
-    return lb->second.endpoints;
+adhoc_t::resolve(const std::string& name) const -> std::shared_ptr<service_description_t> {
+    return m_remotes.apply([&](const remote_map_t& remotes) {
+        remote_map_t::const_iterator lb, ub;
+        std::tie(lb, ub) = remotes.equal_range(name);
+        if(lb == ub) {
+            throw error_t(error::service_not_available, "service {} not found in gateway", name);
```
Comment (3Hren) [inline src/gateway/adhoc.cpp]: Description is redundant here.

File: src/gateway/adhoc.cpp [hunk 4]
```diff
@@ -42,66 +56,109 @@ adhoc_t::~adhoc_t() {
...
-    COCAINE_LOG_DEBUG(m_log, "providing service using remote actor", attribute_list({
-        {"uuid", lb->second.uuid}
-    }));
-
-    return lb->second.endpoints;
+adhoc_t::resolve(const std::string& name) const -> std::shared_ptr<service_description_t> {
+    return m_remotes.apply([&](const remote_map_t& remotes) {
+        remote_map_t::const_iterator lb, ub;
+        std::tie(lb, ub) = remotes.equal_range(name);
+        if(lb == ub) {
```
Comment (3Hren) [inline src/gateway/adhoc.cpp]: Really, it took about minute to figure out what's going on here. Return previous condition.

File: src/gateway/adhoc.cpp [hunk 5]
```diff
@@ -42,66 +56,109 @@ adhoc_t::~adhoc_t() {
...
+            throw error_t(error::service_not_available, "service {} not found in gateway", name);
+        }
+
+        // try to find local service
+        auto it = std::find_if(lb, ub, [&](const remote_map_t::value_type& value) -> bool {
+            return value.second->uuid().empty();
```
Comment (3Hren) [inline src/gateway/adhoc.cpp]: Local services have no UUID? Why?

File: src/gateway/adhoc.cpp [hunk 6]
```diff
@@ -42,66 +56,109 @@ adhoc_t::~adhoc_t() {
...
+        if(it != ub) {
+            COCAINE_LOG_DEBUG(m_log, "providing service using local actor");
+            return it->second;
+        }
+
+        // roll the dice and choose random one
+        std::uniform_int_distribution<int> distribution(0, std::distance(lb, ub) - 1);
+        std::advance(lb, distribution(m_random_generator));
+
+        COCAINE_LOG_DEBUG(m_log, "providing service using remote actor {}", lb->second->uuid());
```
Comment (3Hren) [inline src/gateway/adhoc.cpp]: Return attribute back.

File: src/gateway/adhoc.cpp [hunk 7]
```diff
@@ -42,66 +56,109 @@ adhoc_t::~adhoc_t() {
...
-    auto ptr = m_remotes.synchronize();
+auto
+adhoc_t::on_remote_service_exposed(const std::string& uuid,
+                                   const std::string& name,
+                                   unsigned int version,
+                                   const std::vector<asio::ip::tcp::endpoint>& endpoints,
+                                   const io::graph_root_t& protocol) -> void {
+    m_remotes.apply([&](remote_map_t& remotes){
+        remotes.insert({name, std::make_shared<remote_t>(remote_t{uuid, version, endpoints, protocol})});
+        COCAINE_LOG_DEBUG(m_log,
```
Comment (3Hren) [inline src/gateway/adhoc.cpp]: Ugly.

File: src/gateway/adhoc.cpp [hunk 8]
```diff
@@ -42,66 +56,109 @@ adhoc_t::~adhoc_t() {
...
-{
-    auto ptr = m_remotes.synchronize();
+auto
+adhoc_t::on_remote_service_exposed(const std::string& uuid,
+                                   const std::string& name,
+                                   unsigned int version,
+                                   const std::vector<asio::ip::tcp::endpoint>& endpoints,
+                                   const io::graph_root_t& protocol) -> void {
+    m_remotes.apply([&](remote_map_t& remotes){
+        remotes.insert({name, std::make_shared<remote_t>(remote_t{uuid, version, endpoints, protocol})});
```
Comment (3Hren) [inline src/gateway/adhoc.cpp]: What if the same service were discovered multiple times?

File: src/gateway/adhoc.cpp [hunk 9]
```diff
@@ -42,66 +56,109 @@ adhoc_t::~adhoc_t() {
...
 
-    auto ptr = m_remotes.synchronize();
+auto
+adhoc_t::on_remote_service_removed(const std::string& uuid, const std::string& name) -> void {
+    m_remotes.apply([&](remote_map_t& remotes){
+        remote_map_t::const_iterator lb, ub;
+        std::tie(lb, ub) = remotes.equal_range(name);
+
+        // Since UUIDs are unique, only one remote will match the specified UUID.
+        auto it = std::find_if(lb, ub, [&](const remote_map_t::value_type& value) -> bool {
```
Comment (3Hren) [inline src/gateway/adhoc.cpp]: BTW `std:: binary_search` fits better.

File: src/gateway/adhoc.cpp [hunk 10]
```diff
@@ -42,66 +56,109 @@ adhoc_t::~adhoc_t() {
...
+            return value.second->uuid() == uuid;
+        });
+        if(it != ub) {
+            COCAINE_LOG_DEBUG(m_log,
```
Comment (3Hren) [inline src/gateway/adhoc.cpp]: Ugly.

File: src/gateway/adhoc.cpp [hunk 11]
```diff
@@ -42,66 +56,109 @@ adhoc_t::~adhoc_t() {
...
+                              "removing {}/{} destination with {:d} endpoints from {}",
+                              name,
+                              it->second->version(),
+                              it->second->endpoints().size(),
+                              uuid);
+            remotes.erase(it);
+        } else {
+            COCAINE_LOG_ERROR(m_log, "destination with uuid {} was not found", uuid);
```
Comment (3Hren) [inline src/gateway/adhoc.cpp]: Stick to the following ERROR logs format: "failed to [WHAT]: [WHY]".

File: src/gateway/adhoc.cpp [hunk 12]
```diff
@@ -42,66 +56,109 @@ adhoc_t::~adhoc_t() {
...
-    auto it = std::find_if(lb, ub, [&](const remote_map_t::value_type& value) -> bool {
-        return value.second.uuid == uuid;
+auto
+adhoc_t::cleanup(const std::string& uuid) -> void {
+    m_remotes.apply([&](remote_map_t& remotes){
+        for(auto it = remotes.begin(); it != remotes.end();) {
+            if(it->second->uuid() == uuid) {
+                auto del_it = it;
+                it++;
+                remotes.erase(del_it);
```
Comment (3Hren) [inline src/gateway/adhoc.cpp]: `it = remotes.erase(it);`

File: src/gateway/adhoc.cpp [hunk 13]
```diff
@@ -42,66 +56,109 @@ adhoc_t::~adhoc_t() {
...
 
-    COCAINE_LOG_DEBUG(m_log, "removing destination with {:d} endpoints", it->second.endpoints.size(), attribute_list({
-        {"service", std::get<0>(name)},
-        {"uuid", uuid                },
-        {"version", (int)std::get<1>(name)}
-    }));
-
-    ptr->erase(it); return ptr->count(name);
+auto
+adhoc_t::total_count(const std::string& partition) -> size_t {
```
Comment (3Hren) [inline src/gateway/adhoc.cpp]: What `partition` means?

File: include/cocaine/api/gateway.hpp (hunk 1 above)
Comment (antmat) [inline include/cocaine/api/gateway.hpp]: Initial idea was to avoid copy. Ok, I'll drop it.

File: src/gateway/adhoc.cpp (hunk 4 above)
Comment (antmat) [inline src/gateway/adhoc.cpp]: It's silly to make 2 lookups consequentially. Iterator comparision is quiet straitforward. I think you do now what
`
auto it = map.find(val);
if(it != map.end()) ...
`
mean

File: src/gateway/adhoc.cpp (hunk 3 above)
Comment (antmat) [inline src/gateway/adhoc.cpp]: ok

File: src/gateway/adhoc.cpp (hunk 5 above)
Comment (antmat) [inline src/gateway/adhoc.cpp]: Local uuids make no sense. We can create separate map though.

File: src/gateway/adhoc.cpp (hunk 6 above)
Comment (antmat) [inline src/gateway/adhoc.cpp]: In fact it's quite hard to read logs in that way.

File: src/gateway/adhoc.cpp (hunk 4 above)
Comment (3Hren) [inline src/gateway/adhoc.cpp]: Convenience and clarity first, performance after.

File: src/gateway/adhoc.cpp (hunk 5 above)
Comment (3Hren) [inline src/gateway/adhoc.cpp]: Maybe it have no sense. Then it'll be better to provide method `is_local` to `remote_t`.

File: src/gateway/adhoc.cpp (hunk 8 above)
Comment (antmat) [inline src/gateway/adhoc.cpp]: It will be there several times. The logic is preserved here and in fact registering same service multiple times is logic violation.

Comment (antmat) [inline src/gateway/adhoc.cpp]: Ok, I'll fix it

**Comment by antmat** [comment]
So, any more comments?

File: src/gateway/adhoc.cpp [hunk 14]
```diff
@@ -42,66 +44,111 @@ adhoc_t::~adhoc_t() {
...
+auto
+adhoc_t::cleanup(const std::string& uuid) -> void {
+    m_remotes.apply([&](remote_map_t& remotes){
+        size_t removed = 0;
+        for(auto it = remotes.begin(); it != remotes.end();) {
+            removed += it->second.erase(uuid);
+            if(it->second.empty()) {
+                auto copy = it;
+                it++;
+                remotes.erase(copy);
```
Comment (3Hren) [inline src/gateway/adhoc.cpp]: Again `it = remotes.erase(copy);`.

File: src/gateway/adhoc.cpp [hunk 15]
```diff
@@ -42,66 +44,111 @@ adhoc_t::~adhoc_t() {
...
-        {"service", std::get<0>(name)},
-        {"uuid", uuid                },
-        {"version", (int)std::get<1>(name)}
-    }));
-
-    ptr->erase(it); return ptr->count(name);
+auto
+adhoc_t::total_count(const std::string& name) -> size_t {
+    return m_remotes.apply([&](remote_map_t& remotes) -> size_t {
+        auto it = remotes.find(name);
```
Comment (3Hren) [inline src/gateway/adhoc.cpp]: `const auto`

File: src/service/locator.cpp [hunk 16]
```diff
@@ -380,30 +341,28 @@ locator_t::locator_t(context_t& context, io_service& asio, const std::string& na
...
-
         m_cluster = m_context.repository().get<api::cluster_t>(type, m_context, *this, name + ":cluster", args);
     }
 
-    if(root.as_object().count("gateway")) {
-        const auto conf = root.as_object().at("gateway").as_object();
-        const auto type = conf.at("type", "unspecified").as_string();
-        const auto args = conf.at("args", dynamic_t::object_t());
+    const auto conf = root.as_object().at("gateway").as_object();
+    const auto type = conf.at("type", "adhoc").as_string();
```
Comment (3Hren) [inline src/service/locator.cpp]: Why did you change default gateway value?

File: src/gateway/adhoc.cpp [hunk 17]
```diff
@@ -42,66 +44,111 @@ adhoc_t::~adhoc_t() {
...
 
-    COCAINE_LOG_DEBUG(m_log, "removing destination with {:d} endpoints", it->second.endpoints.size(), attribute_list({
-        {"service", std::get<0>(name)},
-        {"uuid", uuid                },
-        {"version", (int)std::get<1>(name)}
-    }));
-
-    ptr->erase(it); return ptr->count(name);
+auto
+adhoc_t::total_count(const std::string& name) -> size_t {
```
Comment (3Hren) [inline src/gateway/adhoc.cpp]: Should be const.

File: src/gateway/adhoc.cpp [hunk 18]
```diff
@@ -42,66 +44,111 @@ adhoc_t::~adhoc_t() {
...
 
-    auto ptr = m_remotes.synchronize();
+auto
+adhoc_t::on_remote_service_removed(const std::string& uuid, const std::string& name) -> void {
+    m_remotes.apply([&](remote_map_t& remotes){
+        if(remotes[name].erase(uuid)) {
+            COCAINE_LOG_DEBUG(m_log, "removed service {} provided by {} from gateway", name, uuid);
+        } else {
+            auto str = format("failed to remove service {} provided by {} from gateway: not found", name, uuid);
+            COCAINE_LOG_ERROR(m_log, str);
```
Comment (3Hren) [inline src/gateway/adhoc.cpp]: Drop this. We already have an exception which will be eventually caught and logged.

File: src/gateway/adhoc.cpp [hunk 19]
```diff
@@ -42,66 +44,111 @@ adhoc_t::~adhoc_t() {
...
+        COCAINE_LOG_DEBUG(m_log, "registering {}/{} destination with {:d} endpoints from {}",
+                          name, version, endpoints.size(), uuid);
+
+        auto& services_by_uuid = remotes[name];
+        bool inserted;
+        std::tie(std::ignore, inserted) = services_by_uuid.insert({uuid, remote_t{uuid, version, endpoints, protocol}});
+
+        if(!inserted) {
+            auto str = format("failed to add remote service {} located on {} to gateway: service already registered", name, uuid);
+            COCAINE_LOG_ERROR(m_log, str);
```
Comment (3Hren) [inline src/gateway/adhoc.cpp]: Drop this. We already have an exception which will be eventually caught and logged.

File: src/gateway/adhoc.cpp [hunk 20]
```diff
@@ -42,66 +44,111 @@ adhoc_t::~adhoc_t() {
...
+adhoc_t::on_remote_service_exposed(const std::string& uuid,
+                                   const std::string& name,
+                                   unsigned int version,
+                                   const std::vector<asio::ip::tcp::endpoint>& endpoints,
+                                   const io::graph_root_t& protocol) -> void {
+    m_remotes.apply([&](remote_map_t& remotes){
+        COCAINE_LOG_DEBUG(m_log, "registering {}/{} destination with {:d} endpoints from {}",
+                          name, version, endpoints.size(), uuid);
+
+        auto& services_by_uuid = remotes[name];
```
Comment (3Hren) [inline src/gateway/adhoc.cpp]: Useless variable. Moreover `remotes[name].insert({uuid, remote_t{uuid, version, endpoints, protocol}});` looks better.

File: src/gateway/adhoc.cpp [hunk 21]
```diff
@@ -42,66 +44,111 @@ adhoc_t::~adhoc_t() {
...
-{
-    auto ptr = m_remotes.synchronize();
+auto
+adhoc_t::on_remote_service_exposed(const std::string& uuid,
+                                   const std::string& name,
+                                   unsigned int version,
+                                   const std::vector<asio::ip::tcp::endpoint>& endpoints,
+                                   const io::graph_root_t& protocol) -> void {
+    m_remotes.apply([&](remote_map_t& remotes){
+        COCAINE_LOG_DEBUG(m_log, "registering {}/{} destination with {:d} endpoints from {}",
```
Comment (3Hren) [inline src/gateway/adhoc.cpp]: Again convert pre-logging event to post (after success).

File: src/gateway/adhoc.cpp [hunk 22]
```diff
@@ -42,66 +44,111 @@ adhoc_t::~adhoc_t() {
...
+                          name, version, endpoints.size(), uuid);
+
+        auto& services_by_uuid = remotes[name];
+        bool inserted;
+        std::tie(std::ignore, inserted) = services_by_uuid.insert({uuid, remote_t{uuid, version, endpoints, protocol}});
+
+        if(!inserted) {
+            auto str = format("failed to add remote service {} located on {} to gateway: service already registered", name, uuid);
+            COCAINE_LOG_ERROR(m_log, str);
+            throw error_t(str);
```
Comment (3Hren) [inline src/gateway/adhoc.cpp]: Convert to system_error with error code (not EINVAL).

File: src/gateway/adhoc.cpp [hunk 23]
```diff
@@ -42,66 +44,111 @@ adhoc_t::~adhoc_t() {
...
-    auto ptr = m_remotes.synchronize();
+auto
+adhoc_t::on_remote_service_removed(const std::string& uuid, const std::string& name) -> void {
+    m_remotes.apply([&](remote_map_t& remotes){
+        if(remotes[name].erase(uuid)) {
+            COCAINE_LOG_DEBUG(m_log, "removed service {} provided by {} from gateway", name, uuid);
+        } else {
+            auto str = format("failed to remove service {} provided by {} from gateway: not found", name, uuid);
+            COCAINE_LOG_ERROR(m_log, str);
+            throw error_t(str);
```
Comment (3Hren) [inline src/gateway/adhoc.cpp]: Add error_code.

File: src/gateway/adhoc.cpp [hunk 24]
```diff
@@ -42,66 +44,111 @@ adhoc_t::~adhoc_t() {
...
+adhoc_t::resolve(const std::string& name) const -> service_description_t {
+    return m_remotes.apply([&](const remote_map_t& remotes) {
+        auto by_service_it = remotes.find(name);
+        if(by_service_it == remotes.end() || by_service_it->second.empty()) {
+            throw std::system_error(error::service_not_available);
+        }
+
+        // try to find local service
+        auto& services_by_uuid = by_service_it->second;
+        auto it = services_by_uuid.find({});
```
Comment (3Hren) [inline src/gateway/adhoc.cpp]: Looks really magic. Why'd you ever replaced multimap with two maps?

Comment (antmat) [inline src/gateway/adhoc.cpp]: I can add a constant local_uuid if it helps.
2 maps are easier to use and understand(from my point of view), besides they do not allow duplicate uuid for the same service. Code with multimap with this check looks bulky.

File: src/gateway/adhoc.cpp (hunk 22 above)
Comment (antmat) [inline src/gateway/adhoc.cpp]: In which category should I add it?

Comment (3Hren) [inline src/gateway/adhoc.cpp]: Locator?

Comment (antmat) [inline src/gateway/adhoc.cpp]: I still doubt that we need error_code for this specific case (I don't see any usage for it, and it IS invalid argument passed to function), but ok.

File: include/cocaine/detail/gateway/adhoc.hpp [hunk 25]
```diff
@@ -47,12 +47,16 @@ class adhoc_t:
...
     // TODO: Make sure that remote service metadata is consistent across the whole cluster.
     synchronized<remote_map_t> m_remotes;
-    std::string local_uuid;
 
 public:
     adhoc_t(context_t& context, const std::string& _local_uuid, const std::string& name, const dynamic_t& args);
 
     auto
+    resolve_policy() const ->resolve_policy_t override {
+        return resolve_policy_t::remote_only;
```
Comment (3Hren) [inline include/cocaine/detail/gateway/adhoc.hpp]: So now adhoc won't return local services?

Comment (antmat) [inline include/cocaine/detail/gateway/adhoc.hpp]: yep

Comment (3Hren) [inline include/cocaine/detail/gateway/adhoc.hpp]: Why?

Comment (antmat) [inline include/cocaine/detail/gateway/adhoc.hpp]: because it postulates so by it's policy. Locator tries to resolve local first in this case.
